### ISSUE TITLE
SqlStore Startup: "auto assigned organization not found" but it is most definitely there

### ISSUE THREAD

**Issue by mwasson74** [grafana/grafana#71781]
### What happened?

1. The logs claim that the `"auto assigned organization not found"`
    - I think I have found the corresponding code that should be finding the record [here](https://github.com/grafana/grafana/blob/main/pkg/services/sqlstore/user.go#L153)
2. I deleted the record from the org table
3. I reset the `org_id_seq` back to 1
4. Started grafana: `systemctl start grafana-server`
5. It created the record and started successfully
6. I stopped and then restarted the service
7. Claims again that it cannot find the record even though it just created it

***Full logs here:*** [logs.txt](https://github.com/grafana/grafana/files/12071445/logs.txt)

>logger=migrator t=2023-07-17T09:54:27.898269021-04:00 level=debug msg="Skipping migration: Already executed" id="create table dashboard_public_session"
logger=migrator t=2023-07-17T09:54:27.898279961-04:00 level=info msg="migrations completed" performed=0 skipped=575 duration=5.951451ms
logger=sqlstore t=2023-07-17T09:54:27.899388373-04:00 level=debug msg="Ensuring main org and admin user exist"
logger=sqlstore t=2023-07-17T09:54:27.900570498-04:00 level=debug msg="Creating default org" name="Main Org."
logger=sqlstore t=2023-07-17T09:54:27.901971454-04:00 level=debug msg="auto assigned organization not found"
logger=sqlstore t=2023-07-17T09:54:27.902893163-04:00 level=error msg="failed to insert organization with provided id" org_id=1 err="pq: duplicate key value violates unique constraint \"idx_91707_uqe_org_name\""
logger=sqlstore t=2023-07-17T09:54:27.902925908-04:00 level=info msg="Created default organization"
Error: ✗ pq: Could not complete operation in a failed transaction

### What did you expect to happen?

I expected it to find the default organization record that it created

### Did this work before?

Yes, on 9.4.1, _but_ I did the following to "clean" our server from all of the "wrong" ways colleagues have installed/upgraded Grafana:

1. Removed all files and directories that had to do with Grafana
2. Ran `sudo yum install -y --disablerepo=docker-ce-stable --skip-broken https://dl.grafana.com/enterprise/release/grafana-enterprise-10.0.2-1.x86_64.rpm`
3. Files didn't copy from `/usr/share/grafana/conf/` to `/etc/grafana/` like I expected per the [postinst](https://github.com/grafana/grafana/blob/main/packaging/rpm/control/postinst#L49) script
4. Posted a [question](https://community.grafana.com/t/how-to-force-a-fresh-installation/91487/1) in community.grafana and davidallen5 suggested I try `yum remove grafana`
5. Ran `yum remove grafana`
6. Remnants remained so I removed all manually
7. Ran `sudo yum install -y --disablerepo=docker-ce-stable --skip-broken https://dl.grafana.com/enterprise/release/grafana-enterprise-10.0.2-1.x86_64.rpm` again
8. I witnessed log messages from the postinst script as I thought I should and files were copied to where I thought they should be copied to
9. I was able to start grafana using all defaults
10. I uncommented appropriate lines in `/etc/grafana/grafana.ini` to reflect what we had uncommented in our "old" version of grafana.ini
    - This includes our pre-existing PostgreSQL db
11. I had various issues related to us migrating from SQLite to PostgreSQL where I had to change the type of some `bigint` columns to `bool`
12. Grafana saying that it can't find the auto assigned organization is my last issue

### How do we reproduce it?

1. I wouldn't recommend it but follow the steps in the ***Did this work before?*** section.

### Environment (with versions)?

Grafana: grafana-enterprise-10.0.2
OS: Red Hat Enterprise Linux Server release 7.9 (Maipo)
Browser: N/A here but eventually will use Chrome 114.0.5735.110

### Grafana platform?

A package manager (APT, YUM, BREW, etc.)

**Comment by nightmared** [comment]
Hello,
I believe I have the same problem: when I install and start grafana for the first time, it works successfully; but it fails on ulteriors restarts. I consistently get the same error as the creator of the issue.

I use grafana-10.0.0-1.x86_64 on RHEL8, with a postgresql backend.

I get `auto assigned organization not found`, hence it tries to create a new organization, yet it already exists, so the insertion fails, the postgresql transaction is rolled back, which makes grafana crash instantly.

The core problem is that the organisation does exist but grafana belive it does not.
The fact that it fails because of the transaction rollback event though the code is set to ignore the failure in https://github.com/grafana/grafana/blob/039ed7a5bdcb4bd62b25d52b4348953808c2b63a/pkg/services/sqlstore/user.go#L174 is merely an unpleasant side effect, not the crux of the issue at hand.

I believe that this behaviour is a regression that stems from this commit: https://github.com/grafana/grafana/commit/3e85c9075993dcf47994eb75a6fbf219ff57f2e5

Indeed, the organisation is not found because it performs the following query:
`SELECT "id", ..., "updated" from org WHERE (id=$1) AND CREATED=$2 AND updated=$3 LIMIT 1` , using the current time as the secodn and third arguments.
I doubt this is the intended behaviour for this check, as only the organisation `id` should be used to detect if it exists.

As the fields `Created` and `Updated` as instantiated sooner in the lines https://github.com/grafana/grafana/commit/3e85c9075993dcf47994eb75a6fbf219ff57f2e5#diff-c6399cb5035dadc733601c3694ed48447d6abceab7cf2ffc288689f241e41f08R149-R150, the ORM uses their timestamps when checking for the existence of the org id, thus the check will always fail.

Could someone with more knowledge of grafana internals confirm that this *is not* the expected behaviour?

Thanks

**Comment by nightmared** [comment]
@mellieA Do you need further information to reproduce this bug?

**Comment by mildwonkey** [comment]
Hi @mwasson74 and @nightmared - I'm sorry you're running into this unexpected behavior. I can confirm this is a bug. 


The interesting error (to me) is the duplicate org error:

```
ERROR[08-23|09:46:43] failed to insert organization with provided id logger=sqlstore org_id=1 err="pq: duplicate key value violates unique constraint \"org_pkey\""
```

I was able to reproduce this error by truncating the `user` table but leaving the org table in place. This is a confirmed, reproducible bug - @nightmared , I think you are correct in the root cause.

As a workaround (for fresh installations only!), you can drop or truncate the entire user table and org table, and let grafana recreate both. 

I'm sorry for the bad experience - thanks for taking the time to report this issue.

**Comment by nightmared** [comment]
Thanks!
I guess I was running in the situation because I have an automated system installation that populates a RHEL system with a brand new grafana instance, and then we reboot immediately afterwards, without ever logging into the web interface, hence (supposedly) not populating the `users` table.

For the moment, I have switched back to grafana 9.5, which do not exhibit this behavior, so this is not a blocking issue (on my part, at least).
Do not hesitate to ping me if you want me to test a fix.

**Comment by mildwonkey** [comment]
This patch seems to work for me locally, but I'm in the middle of several other issues right now and don't trust my test, so I'm going to drop it here as a reminder (for me!) to follow up:

```
diff --git a/pkg/services/sqlstore/user.go b/pkg/services/sqlstore/user.go
index fd662fe85a..df029c06d1 100644
--- a/pkg/services/sqlstore/user.go
+++ b/pkg/services/sqlstore/user.go
@@ -146,8 +146,6 @@ func verifyExistingOrg(sess *DBSession, orgId int64) error {

 func (ss *SQLStore) getOrCreateOrg(sess *DBSession, orgName string) (int64, error) {
        var org org.Org
-       org.Created = time.Now()
-       org.Updated = org.Created

        if ss.Cfg.AutoAssignOrg {
                has, err := sess.Where("id=?", ss.Cfg.AutoAssignOrgId).Get(&org)
@@ -177,6 +175,8 @@ func (ss *SQLStore) getOrCreateOrg(sess *DBSession, orgName string) (int64, erro
                        return 0, err
                }
        } else {
+               org.Created = time.Now()
+               org.Updated = org.Created
                org.Name = orgName
                if _, err := sess.InsertOne(&org); err != nil {
                        return 0, err
```

**Comment by mildwonkey** [comment]
> I guess I was running in the situation because I have an automated system installation that populates a RHEL system with a brand new grafana instance, and then we reboot immediately afterwards, without ever logging into the web interface, hence (supposedly) not populating the users table.

That's odd and concerning! The initial org and user are created together when Grafana starts, whether or not you log in, so I'm not sure how you got into that state from a fresh installation. I tried it out on a Centos8 Stream VM (using a postgres database), restarting Grafana a couple of times before loading the UI, and didn't see this issue. If you have any more information or any other suggestions for reproducing what you ran into I'd appreciate it.

**Comment by nightmared** [comment]
Oh, looks like you're right, this is because we actually set `security.disable_initial_admin_creation` in our setup (we use LDAP authentication, so we don't need the default administrative user).

This explain our problem. Thanks, I never would have guessed it out withotu your help!

**Comment by nightmared** [comment]
Yes, because grafana 10.2 (which includes the fix) has not been released yet.

I believe 10.2 should be out in less that a month, if the usual release cadence stays its course.
